### PR TITLE
Produce better error for members in top-level main

### PR DESCRIPTION
**Pull request by Youssef1313**
Reported on Discord by @ZacharyPatten

### CODE REVIEW

**Comment by davidwengier** [comment]
Context:
image

**Comment by Youssef1313** [comment]
The bot didn't like to label my PR. Can someone label it for better tracking? Thanks!

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 1]
```diff
@@ -4617,6 +4617,18 @@ SyntaxKind.FileScopedNamespaceDeclaration or
 
             AddInitializers(builder.InstanceInitializers, instanceInitializers);
             AddInitializers(builder.StaticInitializers, staticInitializers);
+
+            // local functions
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:4621]: Nit: Comment isn't necessary (I think the IDE team normally does this, but we don't).

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 2]
```diff
@@ -6986,4 +6986,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_LineContainsDifferentWhitespace" xml:space="preserve">
     <value>Line contains different whitespace than the closing line of the raw string literal: '{0}' versus '{1}'</value>
   </data>
+  <data name="ERR_UnexpectedMemberInTopLevelMain" xml:space="preserve">
+    <value>A top-level entry point cannot directly contain members such as fields, properties, and methods</value>
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:6990]: The `methods` here feels a bit misleading. It can contain methods, namely local functions. Perhaps `instance methods` would be more clear? I'm open to other wordings as well.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 2)

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 2 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:6990]: `instance methods` is confusing, I think. It could be interpreted as "static methods are okay". We may use the terminology from `MethodKind` enum and say `ordinary methods`, do you think this is good and less confusing?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 1 above)
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:4621]:
```suggestion
```

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:6990]: Ordinary methods sounds good to me.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 3]
```diff
@@ -31992,21 +31992,21 @@ public static int TakeOutParam<T>(T y, out T x)
                 var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular9);
 
                 compilation.VerifyDiagnostics(
-                    // (3,12): error CS0116: A namespace cannot directly contain members such as fields or methods
+                    // (3,12): error CS9010: A top-level entry point cannot directly contain members such as fields, properties, and methods
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:31995]: >A top-level entry point cannot directly contain members such as fields, properties, and methods [](http://example.com/codeflow?start=45&length=95)

My preference would be to not do anything, or to simply adjust the wording for the original error to something like: "A namespace **or a top-level entry point** cannot directly contain members such as fields or methods"

**Comment by AlekseyTs** [comment]
Do we have a test for a scenario that has top level statements, then  a namespace declaration, then a member declaration (outside of the preceding namespace declaration)? Is that member declaration in a top-level entry point? Is the namespace declaration in a top-level entry point?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:31995]: @AlekseyTs Adjusting the existing message sounds reasonable, but won't play nicely with @333fred's suggestion: https://github.com/dotnet/roslyn/pull/59321#discussion_r806361327

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:31995]: > Adjusting the existing message sounds reasonable, but won't play nicely with @333fred's suggestion

I am not sure we need to worry about that because the message clearly talks about **members**. Local functions are not members. Alternatively we can replace methods with properties: "A namespace **or a top-level entry point** cannot directly contain members such as fields or **properties**, etc." Or we could say ordinary methods: "A namespace **or a top-level entry point** cannot directly contain members such as fields or **ordinary** methods, etc."

Bottom line, I prefer us not touching implementation.

**Comment by Youssef1313** [comment]
Replaced by https://github.com/dotnet/roslyn/pull/59734.
